### PR TITLE
ISIC2020 Database Binary Classification Challenge Using a Siamese Network

### PR DESCRIPTION
**Pull request by SebastianDobson**
This Pull Request contributes a Siamese Network Binary Classification Model to the Recognition repository, that classifies images from the ISIC2020 database as either Malignant or Benign skin lesions. More information is provided in the README file. 

------------------------------------------------------------------------------------------------------------------------------
Files:
- assets/ : Images used in the README file.
- README.md : Extensive documentation for the architecture, training process, testing process, results analysis, reproduction instructions, and references for the project. 
- dataset.py : Loads the data, preprocesses it, and pairs images to be used by the model.
- environment.yaml : The required dependencies for the Conda Environment that can be installed to locally run the project algorithm. 
- modules.py : Main Siamese Network model components including the ResNet18 Backbone Feature Extractor, and the MLP Classifier.
- predict.py : The testing process that tests the similarity between unseen images to determine the effectiveness of the model.
- to_subdirectories.py : A helper module to split the imported data into a subdirectory for each class (before manual splitting).
- train.py : The training process for the model which includes both a training and validation step for each epoch,
as well as saving the best model's weights, and plotting the accuracy and loss of two steps for each epoch.

### CODE REVIEW

**Comment by hanemma7moud** [comment]
### This is an initial inspection, no action is required at this point
Recognition Problem : total : **14.5**

Solves problem: This approach lacks direct classification capability for individual images. (**2**)

Implementation functions as intended (**1**)

Good design: good-designed (**0.5**)

Commenting: Clear and sufficient comments throughout the code. (**1**)

Difficulty: Hard (**10**)

Note:
- No use of triplet or contrastive loss, instead the model relies on BCE loss for training the Siamese network.
- Pre-trained Resnet18
- Embedding visualizations are not provided which limits insight into feature separability between classes.
- Class imbalance is partially handled through data augmentation, stratified splitting, and random pairing. However, these methods may be insufficient given the large imbalance ratio.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Adequate design and implementation. </td><td>-1</td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Lacks direct classification capability for individual images.</td><td>-2</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>Data leakage found. Has data splits. Imbalance only partially handled.</td><td>-2</td></tr>
	<tr><td></td><td>Difficulty : Hard.  Siamese network (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. </td><td>-1</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time.</td><td></td></tr>
	<tr><td></td><td>Feedback action required: Feedback marks possible +2 if the requested changes are made. Cache files needs to be removed for merge.</td><td>-2</td></tr>
	<tr><td></td><td>Request Description is good. 1</td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-9</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by gayanku** [comment]
Feedback marks possible +2 if the requested changes are made (see above).

**Comment by SebastianDobson** [comment]
Hi I implemented the feedback but didn't get the extra 2 marks?

**Comment by shakes76** [comment]
Doesnt implement the classification -3 additionally

**Comment by shakes76** [comment]
Cant merge because of conflicting changes to main repo files. Please update for merge, doesn't affect grade.

**Comment by shakes76** [comment]
Feedback applied +2
